File: src/programmers/hash/Programmers2.java
package programmers.hash;

import java.util.HashSet;

public class Programmers2 {
    private boolean solution(String[] phone_book) {
        HashSet<String> hashSet = new HashSet<>();

        boolean answer = true;
        loop: for (int i = 0; i < phone_book.length; i++) {
            String s = phone_book[i];

            for (int j = 1; j < s.length(); j++){
                hashSet.add(s.substring(0, j));
            }

            for (int j = 0; j < phone_book.length; j++) {
                if (i == j) continue;

                if(hashSet.contains(phone_book[j])){
                    answer = false;
                    break loop;
                }
            }
        }

        return answer;
    }

    public static void main(String[] args) {
        System.out.println(new Programmers2().solution(new String[]{"119", "97674223", "1195524421"}));
    }
}

File: src/programmers/stack/Programmers2.kt
package programmers.stack

import java.util.*

/**
 * 레이저와 쇠막대 문제
 */
class Programmers2 {
    fun solution(arrangement: String): Int {
        val stack = Stack<Char>()
        var answer = 0

        arrangement.forEachIndexed { index, c ->
            if (c == '(') {
                stack.push(c)
            } else {
                stack.pop()

                if (arrangement[index - 1] != '(') {
                    answer++
                } else {
                    answer += stack.size
                }
            }
        }

        return answer
    }
}

fun main(args: Array<String>) {
    println(
        Programmers2().solution("(()()())")
    )
}